chain/types/ethtypes/eth_transactions.go

```diff
@@ -103,13 +121,79 @@ func NewEthTxArgsFromMessage(msg *types.Message) (EthTxArgs, error) {
 		Nonce:                int(msg.Nonce),
 		To:                   to,
 		Value:                msg.Value,
-		Input:                decodedParams,
+		Input:                params,
 		MaxFeePerGas:         msg.GasFeeCap,
 		MaxPriorityFeePerGas: msg.GasPremium,
 		GasLimit:             int(msg.GasLimit),
 	}, nil
 }
 
+func MessageFromEthTxArgsAndSender(args EthTxArgs, from address.Address) (*types.Message, error) {
+	if args.ChainID != build.Eip155ChainId {
+		return nil, xerrors.Errorf("unsupported chain id: %d", args.ChainID)
+	}
+
+	if !args.V.NilOrZero() || !args.R.NilOrZero() || !args.S.NilOrZero() {
+		return nil, xerrors.New("tx signature fields should be 0")
+	}
+
+	var err error
+	method := builtintypes.MethodSend
+	var params []byte
+	to := address.Undef
+	// nil indicates the EAM, only CreateExternal is allowed
+	if args.To == nil {
+		to = builtintypes.EthereumAddressManagerActorAddr
+		// TODO: Uncomment
+		//method = builtintypes.MethodsEAM.CreateExternal
+		method = builtintypes.MethodsEAM.Create
+		if len(args.Input) == 0 {
+			return nil, xerrors.New("cannot call CreateExternal without params")
+		}
+		// TODO: CreateExternalParams, it doesn't have a nonce
+		params, err = actors.SerializeParams(&eam.CreateParams{
+			Initcode: args.Input,
+			Nonce:    uint64(args.Nonce),
+		})
+		if err != nil {
+			return nil, fmt.Errorf("failed to serialize Create params: %w", err)
+		}
+
+	} else {
+		to, err = args.To.ToFilecoinAddress()
+		if err != nil {
+			return nil, xerrors.Errorf("failed to convert To into filecoin addr: %w", err)
+		}
+		if len(args.Input) == 0 {
+			// Yes, this is redundant, but let's be sure what we're doing
+			method = builtintypes.MethodSend
+			params = make([]byte, 0)
+		} else {
+			// must be InvokeContract
+			method = builtintypes.MethodsEVM.InvokeContract
+			buf := new(bytes.Buffer)
+			if err = cbg.WriteByteArray(buf, args.Input); err != nil {
+				return nil, xerrors.Errorf("failed to write input args: %w", err)
+			}
+
+			params = buf.Bytes()
+		}
+	}
+
+	return &types.Message{
+		Version:    0,
+		To:         to,
+		From:       from,
+		Nonce:      uint64(args.Nonce),
+		Value:      args.Value,
+		GasLimit:   int64(args.GasLimit),
+		GasFeeCap:  args.MaxFeePerGas,
+		GasPremium: args.MaxPriorityFeePerGas,
+		Method:     method,
+		Params:     params,
+	}, nil
+}
+
 func (tx *EthTxArgs) ToSignedMessage() (*types.SignedMessage, error) {
```

Comment:
Oops, I didn't notice the method right after this one does the same thing. We'll need to either consolidate these methods or drop the new one before merging.